### PR TITLE
pgvector 0.7.0 blog post

### PR DESCRIPTION
**Pull request by olirice**
Adds pgvector blog post

### CODE REVIEW

File: apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+Real-world embedding datasets contain a lot of redundancy. It’s buried in the vector space. For example, when vectors are clustered around some central points in multidimensional, we’re seeing exploitable structure. Reducing this makes it possible to get memory and performance savings with no effect on precision. There are several different approaches to employ this ideas made available in pgvector since 0.7.0:
+
+- float16 vector representation
+- sparse vectors
+- bit vectors
+
+### Float16 vectors
+
+An HNSW index is are most efficient when the index fits into shared memory and is not evicted due to concurrent Postgres operations. The reason for this is that it allows Postgres to minimize costly IO. Historically, pgvector supported only 32-bit vectors. In version 0.7.0 pgvector introduces 16-bit float HNSW indexes which consume exactly half the memory. That reduction in memory keeps operations at maximum performance for twice as long.
```
Comment (kevcodez) [inline apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx:26]:
```suggestion
An HNSW index is most efficient when the index fits into shared memory and is not evicted due to concurrent Postgres operations. The reason for this is that it allows Postgres to minimize costly IO. Historically, pgvector supported only 32-bit vectors. In version 0.7.0 pgvector introduces 16-bit float HNSW indexes which consume exactly half the memory. That reduction in memory keeps operations at maximum performance for twice as long.
```

Comment (gregnr) [inline apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx:26]:
```suggestion
An HNSW index is most efficient when it fits into shared memory and avoids being evicted due to concurrent operations, which Postgres performs to minimize costly I/O operations. Historically, pgvector supported only 32-bit vectors. In version 0.7.0, pgvector introduces 16-bit float HNSW indexes which consume exactly half the memory. That reduction in memory keeps operations at maximum performance for twice as long.
```

File: apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+
+There are two options when using float16 vectors:
+
+- Index using float16, but the underlying table continues to use float32
+- The index and the underlying table both use float16. This options uses 50% as much disk space in addition to requiring 50% less shared memory to operate efficiently. Performance is further improved with more vectors fitting in a single Postgres page and with fewer page evictions due to concurrent operations..
```
Comment (gregnr) [inline apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx:31]:
```suggestion
- The index and the underlying table both use float16. This options uses 50% as much disk space in addition to requiring 50% less shared memory to operate efficiently. Performance is further improved with more vectors fitting in a single Postgres page and with fewer page evictions due to concurrent operations.
```

File: apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+To duplicate existing float32 embedding table to float16 one:
```
Comment (gregnr) [inline apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx:33]:
```suggestion
To duplicate an existing float32 embedding table to float16 one:
```

File: apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+    vector halfvec(1536),
+    primary key (id)
+);
+
+insert into embedding_half (vector)
+select
+	vector::halfvec(1536)
+from
+	embedding_full;
+```
```
Comment (gregnr) [inline apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx:47]: Consistent indentation:

```suggestion
```sql
create table embedding_half (
  id serial,
  vector halfvec(1536),
  primary key (id)
);

insert into embedding_half (vector)
select
  vector::halfvec(1536)
from
  embedding_full;
```
```

File: apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+
+With 900K OpenAI 1536-dimensional vectors the table size is 3.5Gb. For comparison, embedding_full required 7Gb.
```
Comment (gregnr) [inline apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx:49]:
```suggestion
With 900K OpenAI 1536-dimensional vectors, the table size is 3.5Gb. For comparison, `embedding_full` required 7Gb.
```

File: apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+
+Then we can build a float16 HNSW index :
```
Comment (gregnr) [inline apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx:51]:
```suggestion
Then we can build a float16 HNSW index:
```

File: apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+
+```sql
+CREATE INDEX ON embedding_half USING hnsw (vector halfvec_l2_ops);
```
Comment (gregnr) [inline apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx:54]: Consistent casing:

```suggestion
create index on embedding_half using hnsw (vector halfvec_l2_ops);
```

File: apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+```
+
+For testing index creation performance we chose a c7g.metal instance with 128Gb memory and parameters:
```
Comment (gregnr) [inline apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx:57]:
```suggestion
To test the performance of index creation, we chose a `c7g.metal` instance with 128Gb memory and the following parameters:
```

File: apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+<Img
+  alt="security-peformance-advisor"
+  src="/images/blog/pgvector-0-7-0-perf.png"
+  captionAlign="left"
+  zoomable={false}
+/>
+
+HNSW build times recently experienced a stepwise improvement in the 0.6.2 release, which introduced parallel builds. 0.7.0 with the halfvec (float16) feature improves that speedup a further 30%.
+
+Note that float16 vector arithmetic on ARM architecture is identical to float32, so that serial build time (with one parallel worker) has not improved. Still there is a significant difference for parallel build due to better pages/IO utilization. I haven’t used pre-warming and other artificial things in this test. Both heap and HNSW relations for float16 occupy only half of the space compared to the previous float32 ones.
```
Comment (gregnr) [inline apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx:79]:
```suggestion
Note that float16 vector arithmetic on the ARM architecture is identical to float32, so serial build times (with one parallel worker) have not improved. However there is a significant difference for parallel builds due to better pages and I/O utilization. Also note that this test doesn't use pre-warming or other artificial enhancements.

Both heap and HNSW relations for float16 occupy only half of the space compared to the previous float32 ones.
```

File: apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+
+There is a proposal to speed it up even more in the future by using SVE intrinsic on ARM architecture (link: https://github.com/pgvector/pgvector/pull/536).
```
Comment (gregnr) [inline apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx:81]:
```suggestion
There is a proposal to speed it up even more in the future by using SVE intrinsics on ARM architecture (see: https://github.com/pgvector/pgvector/pull/536).
```

File: apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+
+Jonathan Katz [made his measurements](https://jkatz05.com/post/postgres/pgvector-performance-150x-speedup) on HNSW performance on r7gd.16xlarge (64 vCPU, 512GiB RAM), and his results are even better. For float16, HNSW build time is up to 3x faster. For select’s performance, ANN benchmark results show that precision is not changed with decreasing bitness, and queries per second (QPS) is similar for in-memory cases. But when in real machine queries are using IO or there is some HNSW pages eviction from memory due to concurrent connections, there would be a meaningful difference. With only half of memory needed to accommodate the same HNSW index, cost for the same performance and precision is also significantly less.
```
Comment (gregnr) [inline apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx:83]:
```suggestion
Jonathan Katz [made his measurements](https://jkatz05.com/post/postgres/pgvector-performance-150x-speedup) on HNSW performance using `r7gd.16xlarge` (64 vCPU, 512GiB RAM), and his results are even better. For float16, HNSW build times are up to 3x faster. For select’s performance, ANN benchmark results show that precision is not changed with decreasing bitness, and queries per second (QPS) is similar to in-memory cases. But when real machine queries are using I/O or some HNSW pages are evicted from memory due to concurrent connections, there would be a meaningful difference. With only half of memory needed to accommodate the same HNSW index, cost for the same performance and precision is also significantly less.
```

File: apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+| **Index size (MB)**        | 7734            | 3867             |
+| **Index build time (s)**   | 264             | 90               |
+| **Recall @ ef_search=10**  | 0.819           | 0.809            |
+| **QPS @ ef_search=10**     | 1231            | 1219             |
+| **Recall @ ef_search=40**  | 0.945           | 0.945            |
+| **QPS @ ef_search=40**     | 627             | 642              |
+| **Recall @ ef_search=200** | 0.987           | 0.987            |
+| **QPS @ ef_search=200**    | 191             | 190              |
+
+Full results on different datasets see [this GitHub issue](https://github.com/pgvector/pgvector/issues/326#issuecomment-2028467121).
```
Comment (gregnr) [inline apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx:96]:
```suggestion
For full results on the different datasets, see [this GitHub issue](https://github.com/pgvector/pgvector/issues/326#issuecomment-2028467121).
```

File: apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx [hunk 13]
```diff
@@ -0,0 +1,184 @@
...
+If vectors contain many zero components, then a sparse vector representation can save significant storage space. For example, to populate sparse vectors:
+
+```sql
+create embedding_sparse (
+    id serial,
+    vector sparsevec(1536),
+    primary key (id)
+)
+
+INSERT INTO embedding_sparse (embedding) VALUES ('{1:0.1,3:0.2,5:0.3}/1536'), ('{1:0.4,3:0.5,5:0.6}/1536');
```
Comment (gregnr) [inline apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx:109]: Consistent indentation/casing:

```suggestion
create embedding_sparse (
  id serial,
  vector sparsevec(1536),
  primary key (id)
)

insert into embedding_sparse (embedding) values ('{1:0.1,3:0.2,5:0.3}/1536'), ('{1:0.4,3:0.5,5:0.6}/1536');
```

File: apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx [hunk 14]
```diff
@@ -0,0 +1,184 @@
...
+```
+
+The sparse vector only consumes storage space required non-zero components. In this case, thats 3 values in a 1536 vector.
+
+Note the new vector syntax `{1:3,3:1,5:2}/1536` for the sparse vector representation in:
+
+```sql
+SELECT * FROM embedding_sparse ORDER BY vector <-> '{1:3,3:1,5:2}/1536' LIMIT 5;
```
Comment (gregnr) [inline apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx:117]:
```suggestion
select * from embedding_sparse order by vector <-> '{1:3,3:1,5:2}/1536' limit 5;
```

File: apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx [hunk 15]
```diff
@@ -0,0 +1,184 @@
...
+create embedding_sparse (
+    id serial,
+    vector sparsevec(1536),
+    primary key (id)
+)
+
+INSERT INTO embedding_sparse (embedding) VALUES ('{1:0.1,3:0.2,5:0.3}/1536'), ('{1:0.4,3:0.5,5:0.6}/1536');
+```
+
+The sparse vector only consumes storage space required non-zero components. In this case, thats 3 values in a 1536 vector.
```
Comment (gregnr) [inline apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx:112]:
```suggestion
The sparse vector only consumes storage space for the non-zero components. In this case, thats 3 values in a 1536 vector.
```

File: apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx [hunk 16]
```diff
@@ -0,0 +1,184 @@
...
+
+Note the new vector syntax `{1:3,3:1,5:2}/1536` for the sparse vector representation in:
+
+```sql
+SELECT * FROM embedding_sparse ORDER BY vector <-> '{1:3,3:1,5:2}/1536' LIMIT 5;
+```
+
+### Bit vectors
+
+Using binary quantization we can represent float vector as a vector in binary space. This reduces storage size dramatically and is intended as a way to quickly “pre-select” from a data set before performaing an additional search within the subset. When properly parameterized, the secondary select can be very fast, even without index.
```
Comment (gregnr) [inline apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx:122]:
```suggestion
Using binary quantization we can represent float vector as a vector in binary space. This reduces storage size dramatically and is intended as a way to quickly “pre-select” from a data set before performing an additional search within the subset. When properly parameterized, the secondary select can be very fast, even without an index.
```

File: apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx [hunk 17]
```diff
@@ -0,0 +1,184 @@
...
+CREATE INDEX ON embedding
+	USING hnsw ((binary_quantize(vector)::bit(1000)) bit_hamming_ops);
+
+SELECT
+	*
+FROM
+	embedding
+ORDER BY
+	binary_quantize(vector)::bit(3) <~> binary_quantize('[1,-2,3]')
+LIMIT 5;
```
Comment (gregnr) [inline apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx:134]:
```suggestion
create index on embedding
  using hnsw ((binary_quantize(vector)::bit(1000)) bit_hamming_ops);

select
  *
from
  embedding
order by
  binary_quantize(vector)::bit(3) <~> binary_quantize('[1,-2,3]')
limit 5;
```

File: apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx [hunk 18]
```diff
@@ -0,0 +1,184 @@
...
+```
+
+To use binary quantized HNSW index to pre-select from a larger dataset and then make a fast selection from the resulting subset, without an index.
```
Comment (gregnr) [inline apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx:137]:
```suggestion
To use a binary quantized HNSW index to pre-select from a larger dataset and then make a fast selection from the resulting subset, without an index:
```

File: apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx [hunk 19]
```diff
@@ -0,0 +1,184 @@
...
+	    *
+    FROM
+	    embedding
+	  ORDER BY
+		  binary_quantize(vector)::bit(3) <~> binary_quantize('[1,-2,3]')
+		LIMIT 20
+)
+ORDER BY
+	vector <=> '[1,-2,3]'
+LIMIT 5;
```
Comment (gregnr) [inline apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx:151]:
```suggestion
select * from (
  select
    *
  from
    embedding
  order by
    binary_quantize(vector)::bit(3) <~> binary_quantize('[1,-2,3]')
  limit 20
)
order by
  vector <=> '[1,-2,3]'
limit 5;
```

File: apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx [hunk 20]
```diff
@@ -0,0 +1,184 @@
...
+```
+
+It allows build a small and fast HNSW index for select, insert or update operations, and still have fast vector search. Exact configuration the `limit` clauses is data dependent, so you’ll want to experiment with the sub-select size and number of final results directly on your own dataset.
```
Comment (gregnr) [inline apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx:154]:
```suggestion
It allows building a small and fast HNSW index for select, insert, or update operations while still having fast vector search. Exact configuration for the `limit` clauses are data dependent, so you’ll want to experiment with the sub-select size and the number of final results directly on your own dataset.
```

File: apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx [hunk 21]
```diff
@@ -0,0 +1,184 @@
...
+### New distance functions
+
+Pgvector 0.7.0 also added support for L1 distance operator `<+>`
+
+New distance types for indexing:
+
+[L1 distance](https://en.wikipedia.org/wiki/Taxicab_geometry) - added in 0.7.0
+
+```sql
+CREATE INDEX ON items USING hnsw (embedding vector_l1_ops);
```
Comment (gregnr) [inline apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx:165]:
```suggestion
create index on items using hnsw (embedding vector_l1_ops);
```

File: apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx [hunk 22]
```diff
@@ -0,0 +1,184 @@
...
+```
+
+[Hamming distance](https://www.notion.so/What-s-new-in-pgvector-0-7-0-fd27d50c8d7e4186a0a4a4bc8b69c7ab?pvs=21) - added in 0.7.0
+
+```sql
+CREATE INDEX ON items USING hnsw (embedding bit_hamming_ops);
```
Comment (gregnr) [inline apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx:171]:
```suggestion
create index on items using hnsw (embedding bit_hamming_ops);
```

File: apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx [hunk 23]
```diff
@@ -0,0 +1,184 @@
...
+```
+
+[Jaccard distance](https://en.wikipedia.org/wiki/Jaccard_index) - added in 0.7.0
+
+```sql
+CREATE INDEX ON vector USING hnsw (vector bit_jaccard_ops);
```
Comment (gregnr) [inline apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx:177]:
```suggestion
create index on vector using hnsw (vector bit_jaccard_ops);
```

File: apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx [hunk 24]
```diff
@@ -0,0 +1,184 @@
...
+ORDER BY
+	vector <=> '[1,-2,3]'
+LIMIT 5;
+```
+
+It allows build a small and fast HNSW index for select, insert or update operations, and still have fast vector search. Exact configuration the `limit` clauses is data dependent, so you’ll want to experiment with the sub-select size and number of final results directly on your own dataset.
+
+### New distance functions
+
+Pgvector 0.7.0 also added support for L1 distance operator `<+>`
```
Comment (gregnr) [inline apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx:158]:
```suggestion
pgvector 0.7.0 also added support for L1 distance operator `<+>`.
```

File: apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx [hunk 25]
```diff
@@ -0,0 +1,184 @@
...
+
+New distance types for indexing:
```
Comment (gregnr) [inline apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx:160]:
```suggestion
And new distance types for indexing:
```

File: apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx [hunk 26]
```diff
@@ -0,0 +1,184 @@
...
+
+[Jaccard distance](https://en.wikipedia.org/wiki/Jaccard_index) - added in 0.7.0
+
+```sql
+CREATE INDEX ON vector USING hnsw (vector bit_jaccard_ops);
+```
+
+### Conclusion
+
+Over the last year pgvector had significant development both is functionality and performance. HNSW, parallel build, and many other options were introduced. In many cases now we have +100 times speedup compared to one year ago.
```
Comment (gregnr) [inline apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx:182]:
```suggestion
Over the last year pgvector has had significant development in both functionality and performance, including HNSW indexes, parallel builds, and many other options. With the introduction of half vectors (float16), sparse vectors, and bit vectors, we're now seeing over 100x speedup compared to one year ago.
```

File: apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx [hunk 27]
```diff
@@ -0,0 +1,184 @@
...
+
+For a more complete comparison of pgvector performance over the last year check out [this post](https://jkatz05.com/post/postgres/pgvector-performance-150x-speedup/).
```
Comment (gregnr) [inline apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx:184]:
```suggestion
For a more complete comparison of pgvector performance over the last year, check out [this post by Jonathan Katz](https://jkatz05.com/post/postgres/pgvector-performance-150x-speedup/).
```

File: apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx (hunk 1 above)
Comment (gregnr) [reply apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx]: Batched this in with some other suggestions

**Comment by gregnr** [comment]
Let's hold off on merging until we add a section explaining when this will be available.

**Review by gregnr — APPROVED** [review]

File: apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx [hunk 28]
```diff
@@ -0,0 +1,188 @@
...
+description: 'Exploring new features in pgvector v0.7.0'
+author: pavel
+image: pgvector-0-7-0.png
+thumb: pgvector-0-7-0.png
+categories:
+  - engineering
+tags:
+  - supabase-engineering
+  - planetpg
+date: '2024-05-01'
```
Comment (awalias) [inline apps/www/_blog/2024-05-01-pgvector-0-7-0.mdx:12]:
```suggestion
date: '2024-05-02'
```

**Review by awalias — APPROVED** [review]

**Pull request merged**